<?xml version="1.0"?>

<launch>

    <!--
    <node pkg="sensor_node_pkg" type="sensor_node_type" name="sensor_mode_name" output="screen">
        <param name="sensor_param" value="param_value" />
    </node>
    -->

    <!-- velodyne sensor configuration and drivers -->

    <include file="$(find velodyne_pointcloud)/launch/VLP16_points.launch" />

    <!-- sbg driver for the ellipse-n imu w/ gnss -->>

    <include file="$(find kart_2dnav)/launch/sbg_ellipseN.launch" />

    <!-- slice up the 3d point cloud into laserscan messages -->

    <node pkg="pointcloud_to_laserscan" type="pointcloud_to_laserscan_node" name="pointcloud_to_laserscan">
        <remap from="cloud_in" to="/velodyne_points" />
        <remap from="scan" to="kart/scan" />
    </node>

    <!--
    <node pkg="kart_setup_odom_tf" type="tf_odom_broadcaster" name="odom_tf" output="screen">
    </node>
    -->

    <!-- perform transform from velodyne sensor to kart base -->

    <!-- <node pkg="kart_setup_lidar_tf" type="tf_broadcaster" name="robot_tf" output="screen">
    </node> -->

    <node pkg="tf" type="static_transform_publisher" name="base_footprint_to_base_link" args="0 0 0 0 0 0 /base_footprint /base_link 100"/>

    <master auto="start"/>

    <!-- Run the robot pose ekf node to combine imu and odom data -->
    <node pkg="robot_pose_ekf" type="robot_pose_ekf" name="robot_pose_ekf">
        <rosparam file="$(find kart_2dnav)/params/robot_pose_ekf_params.yaml" command="load" />
        <remap from="imu_data" to="imu/data" />
    </node>

    <!-- Convert the incoming sbg data to imu data -->
    <node pkg="sbg_to_imu" type="sbg_to_imu.py" name="sbg_to_imu" output="screen" />
    <node pkg="sbg_to_imu" type="fake_odom.py" name="fake_odom" output="screen" />


    <!-- Run the map server -->
    <node pkg="tf" type="static_transform_publisher" name="base_link_to_velodyne" args="0 0 0 0 0 0 /base_link /velodyne 100"/>


    <node pkg="gmapping" type="slam_gmapping" name="gmapping" output="screen">
        <remap from="scan" to="kart/scan" />
        <rosparam file="$(find kart_2dnav)/params/slam_gmapping_params.yaml" command="load" />
    </node>

    <!-- Run AMCL --> 
    
    <node pkg="amcl" type="amcl" name="amcl" output="screen">
        <remap from="scan" to="kart/scan" />
        <rosparam file="$(find kart_2dnav)/params/amcl_params.yaml" command="load" />
    </node>


    <!-- Run the laser_scan_matcher to get odom info from the LiDAR -->

    <!-- <node pkg="laser_scan_matcher" type="laser_scan_matcher_node" name="laser_scan_matcher_node" output="screen">
        <rosparam file="$(find kart_2dnav)/params/laser_scan_matcher_params.yaml" command="load" />
    </node> -->

    <!-- Run the move_base to execute the navigation stack -->

    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
        <rosparam file="$(find kart_2dnav)/params/costmap_common_params.yaml" command="load" ns="global_costmap" /> 
        <rosparam file="$(find kart_2dnav)/params/costmap_common_params.yaml" command="load" ns="local_costmap" />
        <rosparam file="$(find kart_2dnav)/params/local_costmap_params.yaml" command="load" />
        <rosparam file="$(find kart_2dnav)/params/global_costmap_params.yaml" command="load" /> 
        <rosparam file="$(find kart_2dnav)/params/teb_local_planner_params.yaml" command="load" />


        <param name="base_global_planner" value="global_planner/GlobalPlanner" />
        <param name="planner_frequency" value ="1.0" />
        <param name="planner_patience" value="5.0" />

        <param name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS" />
        <param name="controller_frequency" value="5.0" />
        <param name="controller_patience" value="15.0" />

        <param name="clearing_rotation_allowed" value="false" />
    </node>

    <!-- Run the goal setting script to navigate in SLAM mode -->


    <!-- Launch rviz with the configuration file in order to visualize incoming data -->
    <!-- NOTE: Remove for when actually embedded into the Jetson -->
    
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find kart_2dnav)/params/kart_rviz.rviz" />

</launch>
